### PR TITLE
Implemented modify player's name and readiness within lobby #144, inc…

### PR DESCRIPTION
**Pull request by hilalcomak**
…luding the service mentioned in #146.

- Unit tests for Lobby and LobbyManager.
- Integration tests and tests for LobyControllerTest are missing.

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java [hunk 1]
```diff
@@ -25,6 +27,8 @@
     private final String HANNIBAL_URL = "https://sopra-fs22-group-16-client.herokuapp.com?=";
     private final String QR_API_URL = "https://api.qrserver.com/v1/create-qr-code";
 
+    private Set<String> userNames = new HashSet<>();
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:30]: do we need a separate representation of the user names? Cannot we reuse the names of the players of the playerMap?

Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:30]: If we reuse the ones in playerMap, we get O(n) to check if a name is already used, by keeping this set, we get O(1). I agree reusing playerMap might be more readable.

Comment (paolovic) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:30]: I wouldn't separate the names of the players. You could introduce a, for instance, a HashMap<String, Player>.
Furthermore: I just saw, we have a generatePlayer() method in the Lobby! For me, this a perfect candidate to be outsourced into a PlayerFactory.

**Review by apple00juice — COMMENTED** [review]
Do we want to throw ResponseStatusExceptions in the Lobby and LobbyManager? I just ask myself if this couples them to much to the API. If these methods like setUserName get called outside of the context of a API request, the error makes no sense. Wouldn't it be better to throw something like a IlligalArgumentException (maybe there is something more appropriate) and catch them in the service?

Also in the tests, you can remove the public before the test. SonarCloud will mark it as a code smell 😉

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java [hunk 2]
```diff
@@ -100,6 +107,33 @@ public String getInvitationCode() {
         return this.invitationCode;
     }
 
+    @Override
+    public void setUserName(String token, String newName) {
+        if (userNames.contains(newName))
+            throw new ResponseStatusException(HttpStatus.CONFLICT, "Username " + newName + " is already taken.");
```
Comment (apple00juice) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:113]: Do we want to throw ResponseStatusExceptions in the Lobby and LobbyManager? I just ask myself if this couples them to much to the API. If these methods like setUserName get called outside of the context of a API request, the error makes no sense. Wouldn't it be better to throw something like a IlligalArgumentException (maybe there is something more appropriate) and catch them in the service?

Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:113]: or maybe do the check in the lobby service and the setter has as @precondition that it must be unique

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java (hunk 1 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:30]: I removed the set. but didn't implement PlayerFactory.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java (hunk 2 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:113]: I think we can either:
1. Use different exceptions (e.g. DuplicateUsernameException, DuplicateLobbyNameException, etc.).
2. Use exceptions with the actual error specified in REST interface to return to the client (ResponseStatusException).
As for using the ResponseStatusException outside API calls, are there any plans to run any of this code outside of the server so that entry points are not web calls?

Using something else, such as IllegalArgumentException will force us to either duplicate the checks (one in the caller, another in the callee) or group the error codes for REST into a single INTERNAL_SERVER_ERROR. If we need to have different error codes for different reasons (as stated in the REST specification), grouping them should not be a valid option.


Will leave it as is while a decision is made. Just let me know and I'll change this.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java [hunk 3]
```diff
@@ -206,6 +206,11 @@ public Player generatePlayer(){
         return new Player(generatedId, username, token, team);
     }
 
+    // TODO: Only for testing, feel free to reimplement with corresponding story.
+    public void addPlayer(Player player) {
+        playerMap.put(player.getToken(), player);
```
Comment (apple00juice) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:211]: This should use generatePlayer and leave generatePlayer private, but as we want to rewrite it using a Playerfactor it is fine for now.

**Review by apple00juice — APPROVED** [review]
The rest looks good 😄 .
I don't like the responseStatusException in the lobby, but as the other two option we found up to now include testing twice or not testing in the lobby and due to the fact that this works too we can probably leave it like that

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java [hunk 4]
```diff
@@ -71,7 +73,8 @@ public void setVisibility(Visibility visibility) {
 
     @Override
     public Player removePlayer(String token) {
-        return playerMap.remove(token);
+        Player player = playerMap.remove(token);
+        return player;
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:77]: why the change?
Seems unnecessarily verbose to me at first sight

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java [hunk 5]
```diff
@@ -1,9 +1,13 @@
 package ch.uzh.ifi.hase.soprafs22.lobby;
 
+import ch.uzh.ifi.hase.soprafs22.game.Player;
 import ch.uzh.ifi.hase.soprafs22.lobby.enums.Visibility;
 import ch.uzh.ifi.hase.soprafs22.exceptions.SmallestIdNotCreatableException;
 import ch.uzh.ifi.hase.soprafs22.lobby.interfaces.ILobby;
 import ch.uzh.ifi.hase.soprafs22.lobby.interfaces.ILobbyManager;
+import ch.uzh.ifi.hase.soprafs22.rest.dto.PlayerPutDTO;
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java:8]: PlayerPutDTO in the LobbyManager? I am not sure about that...

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/interfaces/ILobbyManager.java [hunk 6]
```diff
@@ -15,4 +16,6 @@
     ILobby createLobby(String name, Visibility visibility) throws SmallestIdNotCreatableException;
 
     void clear();
+
+    void modifyPlayer(String token, Long lobbyId, PlayerPutDTO playerPutDTO);
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/interfaces/ILobbyManager.java:20]: see above

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java [hunk 7]
```diff
@@ -113,6 +116,32 @@ public String getInvitationCode() {
         return this.invitationCode;
     }
 
+    @Override
+    public void setUserName(String token, String newName) {
+        for (Player player: playerMap.values())
+            if (player.getName().equals(newName))
+                throw new ResponseStatusException(HttpStatus.CONFLICT, "Username " + newName + " is already taken.");
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:123]: ResponseStatusException should be thrown in the LobbyService not in the Lobby.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java [hunk 8]
```diff
@@ -113,6 +116,32 @@ public String getInvitationCode() {
...
+        player.setReady(ready);
+        // Here lobby knows if all players are ready and can inform clients through websocket.
+        // Sum of players that are ready:
+        // long playersReady = playerMap.values().stream().filter(Player::isReady).count();
+    }
+
+    private Player getPlayer(String token) {
+        Player player = playerMap.get(token);
+        if (player == null) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "User not found in lobby.");
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:140]: see above

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java [hunk 9]
```diff
@@ -94,8 +98,27 @@ public void clear() {
...
     }
 
+    @Override
+    public void modifyPlayer(String token, Long lobbyId, PlayerPutDTO playerPutDTO) {
+        ILobby lobby = lobbyMap.get(lobbyId);
+        String newName = playerPutDTO.getName();
+        Boolean ready = playerPutDTO.getReady();
+        if (newName != null) {
+            if (isEmpty(newName))
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Username should not be empty.");
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java:108]: ResponseStatusException in the LobbyManager? I'm not sure about that...

**Review by paolovic — CHANGES_REQUESTED** [review]
The code is fine, but not everything is placed where it belongs to imo.
In general, the flow is as follows:
The controller receives the calls resp. sends the calls to the outside world. That's why every conversion to a PutDTO, GetDTO, PostDTO object is done in the controller.
When it receives a call, it typically calls a service (for now we have only the LobbyService). The service is responsible for forwarding these calls to the Lobby, LobbyManager where the magic happens. In case the Lobby or LobbyManager return something wrong, the LobbyService will throw all ResponseStatusExceptions.
Correct me if I'm wrong @apple00juice

**Comment by apple00juice** [comment]
> The code is fine, but not everything is placed where it belongs to imo. In general, the flow is as follows: The controller receives the calls resp. sends the calls to the outside world. That's why every conversion to a PutDTO, GetDTO, PostDTO object is done in the controller. When it receives a call, it typically calls a service (for now we have only the LobbyService). The service is responsible for forwarding these calls to the Lobby, LobbyManager where the magic happens. In case the Lobby or LobbyManager return something wrong, the LobbyService will throw all ResponseStatusExceptions. Correct me if I'm wrong @apple00juice

Yea, thats also how I understood it. Or at least how they did it in the template if I remember correctly

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java (hunk 4 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:77]: Probably during merge, fixed.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java (hunk 7 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:123]: Done. May I ask why the Service and not the Controller?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java (hunk 8 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:140]: Done

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java (hunk 9 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java:108]: Done

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java (hunk 7 above)
Comment (paolovic) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:123]: Separation of Concerns. The controller is the (more or less) interface to the outside world and nothing more. (And that's how they provided the template).
It makes the life also easier for us. Whenever we find some ResponseStatusException, we know, we have to go to the Service and look what's happening there...

Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java]: thanks

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java [hunk 10]
```diff
@@ -94,8 +97,25 @@ public void clear() {
...
+            lobby.setReady(token, ready);
+        }
+    }
+
     @Override
     public Iterator<ILobby> iterator() {
         return lobbyMap.values().iterator();
     }
+
+    private static boolean isEmpty(String s) {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java:118]: now we have exactly the same check once in the lobbyManager and once in the lobbyService

maybe we should create some external helper class who does this check @apple00juice

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/LobbyService.java [hunk 11]
```diff
@@ -144,6 +147,21 @@ public ILobby getLobby(String token, Long lobbyId) {
 
     }
 
+    public void modifyPlayer(String token, Long lobbyId, String newName, Boolean ready) {
+        try {
+            lobbyManager.modifyPlayer(token, lobbyId, newName, ready);
+        }
+        catch (EmptyUsernameException e) {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/service/LobbyService.java:154]: nice!

**Review by paolovic — APPROVED** [review]

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java (hunk 10 above)
Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyManager.java:118]: Yea we could create a static class, that has all these check methods but more general.
